File: lib/data_structure/Heap.rb
class Heap

  def initialize values = []
    @values = []
    values.each{|value| insert value} unless values.empty?
  end

  def insert value
    @values.push value
    sort_up
  end

  def front
    @values.first
  end

  def delete_front
    swap 0, @values.size-1
    @values.pop
    sort_down
  end

  private
  def swap parent, child
    @values[parent], @values[child] = @values[child], @values[parent]
  end

  def sort_down parent = 0
    left_child = 2*parent + 1
    right_child = left_child + 1

    if @values[right_child] &&
      @values[left_child] < @values[right_child] &&
      @values[parent] < @values[right_child]

      swap parent, right_child
      sort_down parent = right_child

    elsif @values[left_child] &&
      @values[parent] < @values[left_child]

      swap parent, left_child
      sort_down parent = left_child
    end
  end

  def sort_up child = @values.size-1
    parent = (child - 1)/2

    if parent >= 0 && @values[child] > @values[parent]
      swap child, parent
      sort_up parent
    end
  end
end

File: public/javascripts/dataStructure/Heap.js
function Heap() {
  this.heap = [];
};

Heap.prototype.insert = function(value) {
  this.heap.push(value);
  this.sortUp();
  console.log("hello");
};

Heap.prototype.front = function() {
  return this.heap[0];
};

Heap.prototype.deleteFront = function() {
  this.swap(0, this.heap.length-1);
  this.heap.pop();
  this.sortDown();
};

Heap.prototype.swap = function(a,b) {
  var temp = this.heap[a];
  this.heap[a] = this.heap[b];
  this.heap[b] = temp;
};

Heap.prototype.sortUp = function() {
  var heapChild = this.heap.length-1;
  var heapParent = Math.floor((heapChild - 1)/2);

  while(heapParent >= 0 &&
    this.heap[heapChild] > this.heap[heapParent]) {

    this.swap(heapChild, heapParent);
    heapChild = heapParent;
    heapParent = Math.floor((heapChild - 1)/2);
  }
};

Heap.prototype.sortDown = function() {
  var heapParent = 0;

  while(heapParent < this.heap.length) {
    var leftChild = 2*heapParent+1;
    var rightChild = leftChild+1;

    if(this.heap[rightChild] !== undefined &&
      this.heap[leftChild] < this.heap[rightChild] &&
      this.heap[heapParent] < this.heap[rightChild]) {

      this.swap(rightChild, heapParent);
      heapParent = rightChild;
    } else if(this.heap[leftChild] !== undefined &&
      this.heap[heapParent] < this.heap[leftChild]) {

      this.swap(heapParent, leftChild);
      heapParent = leftChild;
    } else {
      return;
    }
  }
};
